Sources/SwiftProtobuf/AnyMessageStorage.swift

```diff
@@ -211,13 +257,58 @@ internal class AnyMessageStorage {
         throw BinaryEncodingError.anyTranscodeFailure
       }
     }
+    if !_typeURL.isEmpty {
+      try visitor.visitSingularStringField(value: _typeURL, fieldNumber: 1)
+    }
+    if !_value.isEmpty {
+      try visitor.visitSingularBytesField(value: _value, fieldNumber: 2)
+    }
   }
 }
 
-// Since things are decoded on demand, hashValue and Equality are a little
-// messy.  Message could be equal, but do to how they are currected, we
-// currently end up with different hashValue and equalty could come back
-// false.
+/// Custom handling for Text format.
+extension AnyMessageStorage {
+  func decodeTextFormat(typeURL url: String, decoder: inout TextFormatDecoder) throws {
+    // Decoding the verbose form requires knowing the type.
+    _typeURL = url
+    guard let messageType = Google_Protobuf_Any.messageType(forTypeURL: url) else {
+      // The type wasn't registered, can't parse it.
+      throw TextFormatDecodingError.malformedText
+    }
+    _valueData = nil
+    let terminator = try decoder.scanner.skipObjectStart()
+    var subDecoder = try TextFormatDecoder(messageType: messageType, scanner: decoder.scanner, terminator: terminator)
+    if messageType == Google_Protobuf_Any.self {
+      var any = Google_Protobuf_Any()
+      try any.decodeTextFormat(decoder: &subDecoder)
+      _message = any
+    } else {
+      _message = messageType.init()
+      try _message!.decodeMessage(decoder: &subDecoder)
+    }
+    decoder.scanner = subDecoder.scanner
+    if try decoder.nextFieldNumber() != nil {
+      // Verbose any can never have additional keys.
+      throw TextFormatDecodingError.malformedText
+    }
+  }
+}
+
+/// Hashable and Equatable conformance
```

Comment:
History tends to be easier to follow if changes like this comment aren't rolled into unrelated commits.